### ISSUE TITLE
Ivy ngFor wont over ride variable in scope

### ISSUE THREAD

**Issue by e-davidson** [angular/angular#30677]
# 🐞 bug report

### Affected Package

 The issue is caused by the new Ivy renderer.

### Is this a regression?

 Yes, it is working as expected when ivy isn't enabled.

### Description

the variable name used for the variable in the ngFor loop can no longer be the same as the variable you're using within the ngFor from the component.

## 🔬 Minimal Reproduction

using @angular/cli@8.0.0-rc.4 create a new project.
Add a property to the AppComponent `someClass: any;`

Add the following to the html template of the component.

```html
<div *ngFor='let someClass of someClass.children' >

</div>
```
enable ivy
run `ng build --aot`

It doesn't matter the type of someClass. Let it be any. 
It appears that ivy can't differentiate between the `someClass` in `let someClass` and `someClass` in `someClass.children`.

The code won't compile unless you change the name to something different.
Like this.

```html
<div *ngFor='let someClass2 of someClass.children' >

</div>
```

 https://stackblitz.com/edit/angular-eurwqh

I'm not sure how to get stack blitz to run with the --aot flag. So here's a git repo with the same code.  https://github.com/e-davidson/angular-ivy-issue 

## 🔥 Exception or Error
<pre><code>

ERROR in : Error: Could not resolve [object Object] / undefined
    at Scope.resolve (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:549:23)
    at tcbResolve (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:827:34)
    at C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:755:74
    at astToTypescript (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\expression.js:47:24)
    at astToTypescript (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\expression.js:59:28)
    at Object.astToTypescript (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\expression.js:54:20)
    at tcbExpression (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:755:29)
    at processAttribute (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:802:28)
    at Array.forEach (<anonymous>)
    at tcbGetInputBindingExpressions (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:792:30)
    at TcbDirectiveOp.execute (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:300:28)
    at Scope.executeOp (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:622:26)
    at Scope.render (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:561:22)
    at Object.generateTypeCheckBlock (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_block.js:43:37)
    at TypeCheckFile.addTypeCheckBlock (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\type_check_file.js:47:41)
    at TypeCheckContext.addTemplate (C:\Users\edavidson\src\temp\MatSortUpdateIssue\MatSortUpdateIssue\node_modules\@angular\compiler-cli\src\ngtsc\typecheck\src\context.js:98:36)

</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 8.0.0-rc.4
Node: 12.3.1
OS: win32 x64
Angular: 8.0.0-rc.5
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.800.0-rc.4
@angular-devkit/build-angular     0.800.0-rc.4
@angular-devkit/build-optimizer   0.800.0-rc.4
@angular-devkit/build-webpack     0.800.0-rc.4
@angular-devkit/core              8.0.0-rc.4
@angular-devkit/schematics        8.0.0-rc.4
@angular/cli                      8.0.0-rc.4
@ngtools/json-schema              1.2.0
@ngtools/webpack                  8.0.0-rc.4
@schematics/angular               8.0.0-rc.4
@schematics/update                0.800.0-rc.4
ng-packagr                        5.2.0
rxjs                              6.5.2
typescript                        3.4.5
webpack                           4.30.0
</code></pre>

**Comment by Airblader** [comment]
I'd say this should throw an error so that developers stop shadowing names.

**Comment by JoostK** [comment]
This may be fixed by #30669

**Comment by kara** [comment]
@e-davidson Could you try again with the next release (8.1.0-beta.0)? It should have the fix in #30669

**Comment by e-davidson** [comment]
@kara 
I tried it and it is fixed.
Then you.
